File: phase-03-subscriber-lidar-brake/python/my_py_pkg/my_py_pkg/auto_brake.py
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from geometry_msgs.msg import Twist
from sensor_msgs.msg import PointCloud2
from sensor_msgs_py import point_cloud2


class AutoBrakeNode(Node):
    def __init__(self):
        super().__init__('auto_brake_node')

        self.publisher = self.create_publisher(Twist, 'cmd_vel', 10)

        self.subscription = self.create_subscription(
            PointCloud2,
            'lidar_points',
            self.cloud_callback,
            qos_profile_sensor_data,
        )

        self.get_logger().info('3D Auto Brake Started!')

    def cloud_callback(self, msg: PointCloud2):
        twist = Twist()
        min_forward_distance = 100.0

        points = point_cloud2.read_points(
            msg, field_names=('x', 'y'), skip_nans=True
        )

        for x, y in points:
            if x > 0.0 and abs(y) < 0.2:
                if x < min_forward_distance:
                    min_forward_distance = float(x)

        if min_forward_distance > 1.0:
            twist.linear.x = 0.2
            twist.angular.z = 0.0
            self.get_logger().info(
                f'Clear ahead (Closest: {min_forward_distance:.2f}m). Moving forward...',
                throttle_duration_sec=1.0,
            )
        else:
            twist.linear.x = 0.0
            twist.angular.z = 0.0
            self.get_logger().warn(
                f'Obstacle detected at {min_forward_distance:.2f}m! BRAKING!',
                throttle_duration_sec=1.0,
            )

        self.publisher.publish(twist)


def main(args=None):
    rclpy.init(args=args)
    node = AutoBrakeNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: phase-03-subscriber-lidar-brake/code/my_cpp_pkg/src/auto_brake.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "sensor_msgs/msg/point_cloud2.hpp"
#include "sensor_msgs/point_cloud2_iterator.hpp"

using std::placeholders::_1;

class AutoBrakeNode : public rclcpp::Node
{
public:
    AutoBrakeNode() : Node("auto_brake_node")
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);

        subscriber_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
            "lidar_points", rclcpp::SensorDataQoS(),
            std::bind(&AutoBrakeNode::cloud_callback, this, _1));

        RCLCPP_INFO(this->get_logger(), "3D Auto Brake Started!");
    }

private:
    void cloud_callback(const sensor_msgs::msg::PointCloud2::SharedPtr msg)
    {
        auto twist_msg = geometry_msgs::msg::Twist();
        float min_forward_distance = 100.0f;

        sensor_msgs::PointCloud2ConstIterator<float> iter_x(*msg, "x");
        sensor_msgs::PointCloud2ConstIterator<float> iter_y(*msg, "y");

        for (; iter_x != iter_x.end(); ++iter_x, ++iter_y) {
            float x = *iter_x;
            float y = *iter_y;

            if (x > 0.0f && std::abs(y) < 0.2f) {
                if (x < min_forward_distance) {
                    min_forward_distance = x;
                }
            }
        }

        if (min_forward_distance > 1.0f) {
            twist_msg.linear.x = 0.2;
            twist_msg.angular.z = 0.0;
            RCLCPP_INFO_THROTTLE(
                this->get_logger(), *this->get_clock(), 1000,
                "Clear ahead (Closest: %.2fm). Moving forward...", min_forward_distance);
        } else {
            twist_msg.linear.x = 0.0;
            twist_msg.angular.z = 0.0;
            RCLCPP_WARN_THROTTLE(
                this->get_logger(), *this->get_clock(), 1000,
                "Obstacle detected at %.2fm! BRAKING!", min_forward_distance);
        }

        publisher_->publish(twist_msg);
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr subscriber_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<AutoBrakeNode>());
    rclcpp::shutdown();
    return 0;
}
